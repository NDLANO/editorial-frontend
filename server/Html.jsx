/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { PropTypes } from 'react';
import { renderToString } from 'react-dom/server';
import serialize from 'serialize-javascript';
import Helmet from 'react-helmet';

import config from '../src/config';

const assets = config.isProduction ? require('../htdocs/assets/assets') : require('./developmentAssets'); // eslint-disable-line import/no-unresolved


const Html = (props) => {
  const { lang, className, component, accessToken, state } = props;
  const configWithAccessToken = { ...config, accessToken };
  const content = component ? renderToString(component) : '';
  const head = Helmet.rewind();

  return (
    <html lang={lang} className={className}>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        {head.title.toComponent()}
        {head.meta.toComponent()}
        {head.script.toComponent()}
        { config.isProduction ? <link rel="stylesheet" type="text/css" href={`/assets/${assets['main.css']}`} /> : null}
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:400,600,700,300italic,400,600,700|Signika:400,600,300,700" />
        <link rel="shortcut icon" href={`/assets/${assets['ndla-favicon.png']}`} type="image/x-icon" />
      </head>
      <body>
        <div id="root" dangerouslySetInnerHTML={{ __html: content }} />
        <script dangerouslySetInnerHTML={{ __html: `window.initialState = ${serialize(state)}` }} />
        <script dangerouslySetInnerHTML={{ __html: `window.config = ${serialize(configWithAccessToken)}` }} />
        <script src={`/assets/${assets['main.js']}`} />
        {/* <script type="text/javascript" async src={`https://cdn.mathjax.org/mathjax/2.7-latest/MathJax.js?config=/assets/${assets['mathjaxConfig.js']}`} /> */}
      </body>
    </html>
  );
};

Html.propTypes = {
  lang: PropTypes.string.isRequired,
  accessToken: PropTypes.string.isRequired,
  component: PropTypes.node,
  state: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  className: PropTypes.string.isRequired,
};

export default Html;
