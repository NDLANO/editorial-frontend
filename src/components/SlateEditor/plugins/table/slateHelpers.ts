/**
 * Copyright (c) 2022-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 *
 */

import { Editor, Element, Node } from 'slate';
import { jsx as slatejsx } from 'slate-hyperscript';
import { defaultTableCellBlock } from './defaultBlocks';
import {
  TableBodyElement,
  TableCaptionElement,
  TableCellElement,
  TableElement,
  TableHeadElement,
  TableRowElement,
} from './interfaces';
import {
  TYPE_TABLE,
  TYPE_TABLE_BODY,
  TYPE_TABLE_CAPTION,
  TYPE_TABLE_CELL,
  TYPE_TABLE_HEAD,
  TYPE_TABLE_ROW,
} from './types';

export const isTable = (node?: Node): node is TableElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE;
};

export const isTableCaption = (node?: Node): node is TableCaptionElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_CAPTION;
};
export const isTableHead = (node?: Node): node is TableHeadElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_HEAD;
};

export const isTableBody = (node?: Node): node is TableBodyElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_BODY;
};

export const isTableRow = (node?: Node): node is TableRowElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_ROW;
};

export const isTableCell = (node?: Node): node is TableCellElement => {
  return Element.isElement(node) && node.type === TYPE_TABLE_CELL;
};

export const hasCellAlignOfType = (editor: Editor, type: string) => {
  // For all selected table cells
  for (const [cell] of Editor.nodes<TableCellElement>(editor, {
    match: node => isTableCell(node),
  })) {
    if (cell.data.align === type) {
      return true;
    }
  }
  return false;
};

export const countCells = (row: TableRowElement, stop?: number) => {
  return row.children
    .map(child => {
      if (!isTableCell(child)) {
        return 0;
      }
      return child.data.colspan;
    })
    .slice(0, stop)
    .reduce((a, b) => a + b);
};

export const getTableBodyWidth = (element: TableHeadElement | TableBodyElement) => {
  const firstRow = element.children[0];
  if (isTableRow(firstRow)) {
    return countCells(firstRow);
  }
  return 0;
};

export const getTableBodyHeight = (element: TableHeadElement | TableBodyElement) => {
  return element.children.length;
};

export const createIdenticalRow = (element: TableRowElement) => {
  return slatejsx(
    'element',
    { type: TYPE_TABLE_ROW },
    element.children.map(child => {
      if (isTableCell(child)) {
        return {
          ...defaultTableCellBlock(),
          data: {
            ...child.data,
            rowspan: 1,
          },
        };
      }
      return defaultTableCellBlock();
    }),
  );
};
