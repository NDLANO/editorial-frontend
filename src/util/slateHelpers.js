/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import {
  reduceElementDataAttributes,
  createEmbedProps,
  reduceChildElements,
} from './embedTagHelpers';

export const BLOCK_TAGS = {
  section: 'section',
  blockquote: 'quote',
  details: 'details',
  summary: 'summary',
  pre: 'code',
  h1: 'heading-two',
  h2: 'heading-two',
  h3: 'heading-two',
  h4: 'heading-two',
  h5: 'heading-two',
  h6: 'heading-two',
  br: 'br',
};

export const TABLE_TAGS = {
  table: 'table',
  th: 'table-cell',
  tr: 'table-row',
  td: 'table-cell',
};

const MARK_TAGS = {
  strong: 'bold',
  em: 'italic',
  u: 'underlined',
  code: 'code',
};

const ListText = ({ children }) => children;

export const findNodesByType = (node, type, nodes = []) => {
  if (node.type === type) {
    nodes.push(node);
  } else if (
    node.kind === 'document' ||
    (node.kind === 'block' && node.nodes.size > 0)
  ) {
    node.nodes.forEach(n => findNodesByType(n, type, nodes));
  }
  return nodes;
};

export const toJSON = state => state.toJSON();

export const logState = state => {
  console.log(JSON.stringify(toJSON(state), null, 2)); // eslint-disable-line no-console
};

// TODO: get type of aside in here. Default should be rightAside since that is the only
const getAsideType = el => ({
  type: el.attributes.getNamedItem('data-type')
    ? el.attributes.getNamedItem('data-type').value
    : 'rightAside',
});

const setAsideTag = data => ({
  'data-type': data.get('type') || '',
});

const illegalTextUnderBlocks = ['ol', 'ul'];

/* eslint-disable consistent-return, default-case */
export const textRule = {
  deserialize(el) {
    if (
      el.nodeName.toLowerCase() === '#text' &&
      el.parentNode &&
      illegalTextUnderBlocks.includes(el.parentNode.tagName.toLowerCase())
    ) {
      return null;
    } else if (
      !el.nodeName ||
      el.nodeName.toLowerCase() !== '#text' ||
      (el.parentNode && el.parentNode.tagName.toLowerCase() !== 'section')
    ) {
      return;
    }
    return null;
  },
};
export const divRule = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'div') return;

    if (el.className === 'c-bodybox') {
      return {
        kind: 'block',
        type: 'bodybox',
        nodes: next(el.childNodes),
      };
    } else if (el.dataset.type === 'related-content') {
      return {
        kind: 'block',
        type: 'related',
        isVoid: true,
        data: reduceChildElements(el),
      };
    }
    const childs = next(el.childNodes);
    return {
      kind: 'block',
      type: 'div',
      nodes: childs,
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'div' && object.type !== 'bodybox') return;
    switch (object.type) {
      case 'bodybox':
        return <div className="c-bodybox">{children}</div>;
      default:
        return <div>{children}</div>;
    }
  },
};

export const paragraphRule = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'p') return;
    const parent = el.parentElement
      ? el.parentElement.tagName.toLowerCase()
      : '';

    const type = parent === 'li' ? 'list-text' : 'paragraph';

    return {
      kind: 'block',
      type,
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'paragraph' && object.type !== 'list-text') return;
    if (object.type === 'list-text') {
      return <ListText>{children}</ListText>;
    }
    return <p>{children}</p>;
  },
};

export const listItemRule = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'li') return;
    return {
      kind: 'block',
      type: 'list-item',
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'list-item') return;
    return <li>{children}</li>;
  },
};

export const unorderListRules = {
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'ul') return;
    const type = el.attributes.getNamedItem('data-type');
    const data = { type: type ? type.value : '' };

    if (data.type === 'two-column') {
      return {
        kind: 'block',
        type: 'two-column-list',
        nodes: next(el.childNodes),
        data,
      };
    }

    return {
      kind: 'block',
      type: 'bulleted-list',
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'two-column-list' && object.type !== 'bulleted-list') {
      return;
    }

    if (object.type === 'two-column-list') {
      return <ul data-type="two-column">{children}</ul>;
    }
    return <ul>{children}</ul>;
  },
};

export const orderListRules = {
  // div handling with text in box (bodybox)
  deserialize(el, next) {
    if (el.tagName.toLowerCase() !== 'ol') return;
    const type = el.attributes.getNamedItem('data-type');
    const data = { type: type ? type.value : '' };
    if (data.type === 'letters') {
      return {
        kind: 'block',
        type: 'letter-list',
        nodes: next(el.childNodes),
        data,
      };
    }
    return {
      kind: 'block',
      type: 'numbered-list',
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    if (object.type !== 'numbered-list' && object.type !== 'letter-list')
      return;
    if (object.type === 'letter-list') {
      return <ol data-type="letters">{children}</ol>;
    }
    return <ol>{children}</ol>;
  },
};

export const footnoteRule = {
  deserialize(el) {
    if (!el.tagName.toLowerCase().startsWith('embed')) return;
    const embed = reduceElementDataAttributes(el);
    if (embed.resource !== 'footnote') return;

    return {
      kind: 'inline',
      type: 'footnote',
      nodes: [
        {
          kind: 'text',
          isVoid: true,
          leaves: [
            {
              kind: 'leaf',
              text: '#',
              marks: [],
            },
          ],
        },
      ],
      data: {
        ...embed,
        authors: embed.authors ? embed.authors.split(';') : [],
      },
    };
  },
  serialize(object) {
    if (object.kind !== 'inline') return;
    if (object.type !== 'footnote') return;

    const data = object.data.toJS();
    const props = createEmbedProps({
      ...data,
      authors: data.authors ? data.authors.join(';') : '',
    });
    return <embed {...props} />;
  },
};

export const blockRules = {
  deserialize(el, next) {
    const block = BLOCK_TAGS[el.tagName.toLowerCase()];
    if (!block) return;
    return {
      kind: 'block',
      type: block,
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    switch (object.type) {
      case 'section':
        return <section>{children}</section>;
      case 'bulleted-list':
        return <ul>{children}</ul>;
      case 'heading-one':
        return <h1>{children}</h1>;
      case 'heading-two':
        return <h2>{children}</h2>;
      case 'heading-three':
        return <h3>{children}</h3>;
      case 'heading-four':
        return <h4>{children}</h4>;
      case 'heading-five':
        return <h5>{children}</h5>;
      case 'heading-six':
        return <h6>{children}</h6>;
      case 'quote':
        return <blockquote>{children}</blockquote>;
      case 'details':
        return <details>{children}</details>;
      case 'summary':
        return <summary>{children}</summary>;
      case 'br':
        return <br />;
    }
  },
};

export const tableRules = {
  deserialize(el, next) {
    const tableTag = TABLE_TAGS[el.tagName.toLowerCase()];
    if (!tableTag) return;
    return {
      kind: 'block',
      type: tableTag,
      nodes: next(el.childNodes),
    };
  },
  serialize(object, children) {
    if (object.kind !== 'block') return;
    switch (object.type) {
      case 'table': {
        return (
          <table>
            <thead>{children.slice(0, 1)}</thead>
            <tbody>{children.slice(1)}</tbody>
          </table>
        );
      }
      case 'table-row':
        return <tr>{children}</tr>;
      case 'table-cell':
        if (object.data.get('isHeader')) {
          return <th>{children}</th>;
        }
        return <td>{children}</td>;
    }
  },
};

const relatedRule = {
  serialize(object) {
    if (object.type === 'related') {
      return (
        <div data-type="related-content">
          {object.data.get('nodes') &&
            object.data
              .get('nodes')
              .map(node => <embed {...createEmbedProps(node)} />)}
        </div>
      );
    }
  },
};

const RULES = [
  divRule,
  textRule,
  orderListRules,
  unorderListRules,
  tableRules,
  paragraphRule,
  listItemRule,
  relatedRule,
  {
    // Aside handling
    deserialize(el, next) {
      if (el.tagName.toLowerCase() !== 'aside') return;
      return {
        kind: 'block',
        type: 'aside',
        nodes: next(el.childNodes),
        data: getAsideType(el),
      };
    },
    serialize(object, children) {
      if (object.kind !== 'block') return;
      if (object.type !== 'aside') return;
      return <aside {...setAsideTag(object.data)}>{children}</aside>;
    },
  },
  blockRules,
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName.toLowerCase()];
      if (!mark) return;
      return {
        kind: 'mark',
        type: mark,
        nodes: next(el.childNodes),
      };
    },
    serialize(object, children) {
      if (object.kind !== 'mark') return;
      switch (object.type) {
        case 'bold':
          return <strong>{children}</strong>;
        case 'italic':
          return <em>{children}</em>;
        case 'underlined':
          return <u>{children}</u>;
        case 'superscripted':
          return <sup>{children}</sup>;
      }
    },
  },
  footnoteRule,
  {
    deserialize(el, next) {
      if (el.tagName.toLowerCase() !== 'a') return;
      return {
        kind: 'inline',
        type: 'link',
        data: {
          href: el.href ? el.href : '#',
          target: el.target ? el.target : '',
          rel: el.rel ? el.rel : '',
        },
        nodes: next(el.childNodes),
      };
    },
    serialize(object, children) {
      if (object.kind !== 'inline') return;
      if (object.type !== 'link') return;
      const data = object.data.toJS();

      if (data.resource === 'content-link') {
        return (
          <embed
            data-resource={data.resource}
            data-content-id={data['content-id']}
            data-link-text={object.text}
            data-open-in={data['open-in']}
          />
        );
      }

      return (
        <a
          href={data.href}
          target={data.target}
          rel={data.rel}
          title={object.text}>
          {children}
        </a>
      );
    },
  },
];

const topicArticeEmbedRule = [
  {
    // Embeds handling
    deserialize(el) {
      if (el.tagName.toLowerCase() !== 'embed') return;

      if (el.dateset['data-resource'] === 'related-content') {
        return;
      }
      return {
        kind: 'block',
        type: 'embed',
        data: reduceElementDataAttributes(el),
        isVoid: true,
      };
    },
    serialize(object) {
      if (object.kind !== 'block') return;
      if (object.type !== 'embed') return;
      switch (object.type) {
        case 'embed':
          return <deleteme />;
      }
    },
  },
];

export const learningResourceEmbedRule = [
  {
    deserialize(el) {
      if (!el.tagName.toLowerCase().startsWith('embed')) return;

      const embed = reduceElementDataAttributes(el);
      if (el.dataset['data-resource'] === 'related-content') return;
      if (embed.resource === 'content-link') {
        return {
          kind: 'inline',
          type: 'link',
          data: embed,
          nodes: [
            {
              kind: 'text',
              text: embed['link-text']
                ? embed['link-text']
                : 'Ukjent link tekst',
            },
          ],
          isVoid: false,
        };
      }
      return {
        kind: 'block',
        type: 'embed',
        data: embed,
        isVoid: true,
      };
    },

    serialize(object) {
      if (!object.type || !object.type.startsWith('embed')) return;
      const data = object.data.toJS();
      const props = createEmbedProps(data);

      return <embed {...props} />;
    },
  },
];

export const topicArticeRules = topicArticeEmbedRule.concat(RULES);
export const learningResourceRules = RULES.concat(learningResourceEmbedRule);
