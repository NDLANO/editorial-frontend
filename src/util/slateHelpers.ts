/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Descendant, Node, Element, ElementType } from "slate";

export const findNodesByType = <T extends ElementType>(
  descendants: Descendant[],
  ...types: readonly T[]
): Extract<Element, { type: T }>[] => {
  return descendants
    .flatMap((descendant) => Array.from(Node.elements(descendant)))
    .map(([node]) => node)
    .filter((node): node is Extract<Element, { type: T }> => types.includes(node.type as T));
};
