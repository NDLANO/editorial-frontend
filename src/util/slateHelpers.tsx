/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Descendant, Node, Element } from 'slate';

export const findNodesByType = (descendants: Descendant[], ...type: Element['type'][]) => {
  const ret = descendants
    .flatMap((descendant) => Array.from(Node.elements(descendant)))
    .map(([node]) => node)
    .filter((node) => type.includes(node.type));
  return ret;
};
