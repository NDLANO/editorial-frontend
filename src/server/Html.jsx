/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { renderToString } from 'react-dom/server';
import serialize from 'serialize-javascript';
import Helmet from 'react-helmet';
import config from '../config';
import { GoogleTagMangerScript, GoogleTagMangerNoScript } from './Gtm';

const assets = require(process.env.RAZZLE_ASSETS_MANIFEST); // eslint-disable-line import/no-dynamic-require

const Html = props => {
  const { lang, className, component, state } = props;
  const content = component ? renderToString(component) : '';
  const head = Helmet.rewind();

  return (
    <html lang={lang} className={className}>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <GoogleTagMangerScript />
        {head.title.toComponent()}
        {head.meta.toComponent()}
        {head.script.toComponent()}
        {assets.client && assets.client.css && (
          <link rel="stylesheet" type="text/css" href={assets.client.css} />
        )}
        <link
          rel="stylesheet"
          href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:400,600,700,300italic,400,600,700|Signika:400,600,300,700"
        />
        <link
          rel="shortcut icon"
          href="/static/ndla-favicon.png"
          type="image/x-icon"
        />
      </head>
      <body>
        <GoogleTagMangerNoScript />
        <script
          dangerouslySetInnerHTML={{
            __html: `
            window.dataLayer = window.dataLayer || [];
            window.originalLocation = { originalLocation: document.location.protocol + '//' + document.location.hostname + document.location.pathname + document.location.search };
            window.dataLayer.push(window.originalLocation);`,
          }}
        />
        <div id="root" dangerouslySetInnerHTML={{ __html: content }} />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.initialState = ${serialize(state)}`,
          }}
        />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.config = ${serialize(config)}`,
          }}
        />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.assets = ${serialize(assets)}`,
          }}
        />
        <script
          type="text/javascript"
          src={assets.client.js}
          defer
          crossOrigin={(process.env.NODE_ENV !== 'production').toString()}
        />
        <script
          type="text/x-mathjax-config"
          dangerouslySetInnerHTML={{
            __html: `
  MathJax.Hub.Config({
    jax: ["input/MathML", "output/CommonHTML"],
    extensions: ["mml2jax.js"],
    showMathMenu: false,
    showProcessingMessages: false,
    menuSettings: {
      zoom: "None"
    }
  });
        `,
          }}
        />
        <script
          type="text/javascript"
          defer
          async
          src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js"
        />
      </body>
    </html>
  );
};

Html.propTypes = {
  lang: PropTypes.string.isRequired,
  component: PropTypes.node,
  state: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  className: PropTypes.string.isRequired,
};

export default Html;
