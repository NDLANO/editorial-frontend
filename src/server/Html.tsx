/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ReactElement } from 'react';
import { renderToString } from 'react-dom/server';
import serialize from 'serialize-javascript';
import config from '../config';
import { GoogleTagMangerScript, GoogleTagMangerNoScript } from './Gtm';

const assets = require(process.env.RAZZLE_ASSETS_MANIFEST as string); // eslint-disable-line import/no-dynamic-require

interface Props {
  lang: string;
  className: string;
  component?: ReactElement;
  state?: {
    locale?: string;
  };
}

const Html = (props: Props) => {
  const { lang, className, component, state = {} } = props;

  const content = component ? renderToString(component) : '';

  return (
    <html lang={lang} className={className}>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <meta httpEquiv="X-UA-Compatible" content="IE=edge" />
        <GoogleTagMangerScript />
        {assets['client.css'] && (
          <link rel="stylesheet" type="text/css" href={assets['client.css']} />
        )}
        <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
        <link rel="manifest" href="/site.webmanifest" />
        <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5" />
        <meta name="msapplication-TileColor" content="#da532c" />
        <meta name="theme-color" content="#ffffff" />
      </head>
      <body>
        <GoogleTagMangerNoScript />
        <script
          dangerouslySetInnerHTML={{
            __html: `
            window.dataLayer = window.dataLayer || [];
            window._mtm = window._mtm || [];
            window.originalLocation = { originalLocation: document.location.protocol + '//' + document.location.hostname + document.location.pathname + document.location.search };
            window.dataLayer.push(window.originalLocation);`,
          }}
        />
        <div id="root" dangerouslySetInnerHTML={{ __html: content }} />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.initialState = ${serialize(state)}`,
          }}
        />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.config = ${serialize(config)}`,
          }}
        />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.assets = ${serialize(assets)}`,
          }}
        />
        <script
          type="text/javascript"
          src={assets['client.js']}
          defer
          crossOrigin={(process.env.NODE_ENV !== 'production').toString()}
        />
        <script
          type="text/x-mathjax-config"
          defer
          dangerouslySetInnerHTML={{
            __html: `
              MathJax = {
                chtml:{
                  mathmlSpacing: false
                },
                options: {
                  renderActions: {
                    addMenu: [],
                    checkLoading: []
                  },
                  menuOptions: {
                    settings: {
                      zoom: "None"
                    }
                  }
                }
              };
        `,
          }}
        />
        <script
          type="text/javascript"
          defer
          src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/mml-chtml.js"
        />
      </body>
    </html>
  );
};

export default Html;
