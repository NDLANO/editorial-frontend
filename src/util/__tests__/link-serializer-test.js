/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
/** @jsx h */

/**
 * @jest-environment node
 */

import h from 'slate-hyperscript';
import jsdom from 'jsdom';
import renderer from 'react-test-renderer';
import Html from 'slate-html-serializer';
import { learningResourceRules } from '../slateHelpers';

const serializer = new Html({
  rules: learningResourceRules,
  parseHtml: jsdom.JSDOM.fragment,
});

test('deserialize a link and a content-link in an list', () => {
  const deserialized = serializer.deserialize(
    '<ul><li> <embed data-content-id="504" data-link-text="Yrker i mediebransjen" data-resource="content-link"> </li><li> <a href="https://www.nrk.no/skole/?page=search&amp;q=Jobb%20i%20NRK" rel="noopener noreferrer" target="_blank" title="Yrker i NRK">Yrker i NRK</a> </li></ul>',
  );

  expect(deserialized.toJSON()).toMatchSnapshot();
});
test('deserialize link with spacing before link', () => {
  const deserialized = serializer.deserialize(
    '<p>Eller du kan bruke nettsidene <a href="http://www.nav.no/" rel="noopener noreferrer" target="_blank" title="nav.no">nav.no</a> . Her finnes informasjon om de ulike tjenestene.</p>',
  );
  expect(deserialized.toJSON()).toMatchSnapshot();
});

test('serialize list with a link and text', () => {
  const state = (
    <state>
      <document>
        <block type="bulleted-list">
          <block type="list-item">
            A string of <mark type="bold">bold</mark> in a{' '}
            <inline
              type="link"
              data={{
                href: 'http://slatejs.org',
                rel: 'noopener noreferrer',
                target: '_blank',
              }}>
              Slate
            </inline>{' '}
            editor!
          </block>
          <block type="list-item">
            A string of <mark type="bold">bold</mark> in a editor!
          </block>
        </block>
      </document>
    </state>
  );

  const list = serializer.serialize(state, { render: false });
  expect(renderer.create(list.first()).toJSON()).toMatchSnapshot();
});
