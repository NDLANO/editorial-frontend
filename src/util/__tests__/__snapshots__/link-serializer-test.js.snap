// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserialize a link and a content-link in an list 1`] = `
Object {
  "document": Object {
    "data": Object {},
    "nodes": Array [
      Object {
        "data": Object {},
        "nodes": Array [
          Object {
            "data": Object {},
            "nodes": Array [
              Object {
                "leaves": Array [
                  Object {
                    "marks": Array [],
                    "object": "leaf",
                    "text": " ",
                  },
                ],
                "object": "text",
              },
              Object {
                "data": Object {
                  "content-id": "504",
                  "link-text": "Yrker i mediebransjen",
                  "resource": "content-link",
                },
                "nodes": Array [
                  Object {
                    "leaves": Array [
                      Object {
                        "marks": Array [],
                        "object": "leaf",
                        "text": "Yrker i mediebransjen",
                      },
                    ],
                    "object": "text",
                  },
                ],
                "object": "inline",
                "type": "link",
              },
              Object {
                "leaves": Array [
                  Object {
                    "marks": Array [],
                    "object": "leaf",
                    "text": " ",
                  },
                ],
                "object": "text",
              },
            ],
            "object": "block",
            "type": "list-item",
          },
          Object {
            "data": Object {},
            "nodes": Array [
              Object {
                "leaves": Array [
                  Object {
                    "marks": Array [],
                    "object": "leaf",
                    "text": " ",
                  },
                ],
                "object": "text",
              },
              Object {
                "data": Object {
                  "href": "https://www.nrk.no/skole/?page=search&q=Jobb%20i%20NRK",
                  "rel": "noopener noreferrer",
                  "target": "_blank",
                  "title": "Yrker i NRK",
                },
                "nodes": Array [
                  Object {
                    "leaves": Array [
                      Object {
                        "marks": Array [],
                        "object": "leaf",
                        "text": "Yrker i NRK",
                      },
                    ],
                    "object": "text",
                  },
                ],
                "object": "inline",
                "type": "link",
              },
              Object {
                "leaves": Array [
                  Object {
                    "marks": Array [],
                    "object": "leaf",
                    "text": " ",
                  },
                ],
                "object": "text",
              },
            ],
            "object": "block",
            "type": "list-item",
          },
        ],
        "object": "block",
        "type": "bulleted-list",
      },
    ],
    "object": "document",
  },
  "object": "value",
}
`;

exports[`deserialize link with spacing before link 1`] = `
Object {
  "document": Object {
    "data": Object {},
    "nodes": Array [
      Object {
        "data": Object {},
        "nodes": Array [
          Object {
            "leaves": Array [
              Object {
                "marks": Array [],
                "object": "leaf",
                "text": "Eller du kan bruke nettsidene ",
              },
            ],
            "object": "text",
          },
          Object {
            "data": Object {
              "href": "http://www.nav.no/",
              "rel": "noopener noreferrer",
              "target": "_blank",
              "title": "nav.no",
            },
            "nodes": Array [
              Object {
                "leaves": Array [
                  Object {
                    "marks": Array [],
                    "object": "leaf",
                    "text": "nav.no",
                  },
                ],
                "object": "text",
              },
            ],
            "object": "inline",
            "type": "link",
          },
          Object {
            "leaves": Array [
              Object {
                "marks": Array [],
                "object": "leaf",
                "text": " . Her finnes informasjon om de ulike tjenestene.",
              },
            ],
            "object": "text",
          },
        ],
        "object": "block",
        "type": "paragraph",
      },
    ],
    "object": "document",
  },
  "object": "value",
}
`;

exports[`serialize list with a link and text 1`] = `
<ul>
  <li>
    A string of 
    <strong>
      bold
    </strong>
     in a 
    <a
      href="http://slatejs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Slate
    </a>
     editor!
  </li>
  <li>
    A string of 
    <strong>
      bold
    </strong>
     in a editor!
  </li>
</ul>
`;
